### PR TITLE
Add FittedBox

### LINKED ISSUE #5830: Add a widget that applies an ImageFit type to its child widget
Example use case: have a Text widget that scales its text size depending on the size of its container.

From @Hixie:

>  
using https://docs.flutter.io/flutter/painting/applyImageFit.html that'd be pretty easy
the short answer is, you create a RenderShiftedBox that is somewhat of a cross between RenderTransform and RenderPadding
that takes an ImageFit instead of an EdgeInset or Matrx4
it then has the child lay itself out unbounded, takes its size, passes that size to that function along with the current widget's size
and then figures the the transform you need to render the widget as per the result
that last part is the only tricky part i think
and yeah, we should totally make this a generic widget

### PR DESCRIPTION
**Pull request by abarth**
This widget lets you apply an ImageFit to a child widget.

Fixes #5830

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie

File: packages/flutter/lib/src/rendering/box.dart [hunk 1]
```diff
@@ -209,6 +209,46 @@ class BoxConstraints extends Constraints {
...
   }
 
+  /// Returns a size that attempts to meet the following conditions, in order:
+  ///
+  ///  - The size must satisfy the constraints.
+  ///  - The aspect ratio of the returned size matches the aspect ratio of the
+  ///    given size.
+  ///  - The retruned size is maximal subject to being smaller than the given
+  ///    size.
+  Size constrainSizeAndAttemptToPreseveAspectRatio(Size size) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: `constrainSizeAndAttemptToPreseveAspectRatio` -> `constrainSizeAndAttemptToPreserveAspectRatio`

File: packages/flutter/lib/src/rendering/box.dart [hunk 2]
```diff
@@ -209,6 +209,46 @@ class BoxConstraints extends Constraints {
     return result;
   }
 
+  /// Returns a size that attempts to meet the following conditions, in order:
+  ///
+  ///  - The size must satisfy the constraints.
+  ///  - The aspect ratio of the returned size matches the aspect ratio of the
+  ///    given size.
+  ///  - The retruned size is maximal subject to being smaller than the given
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: retruned -> returned

Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: Also I don't understand bullet three. Do you mean "The returned size as big as possible while still being equal to or smaller than the given size"?

File: packages/flutter/lib/src/rendering/box.dart [hunk 3]
```diff
@@ -209,6 +209,46 @@ class BoxConstraints extends Constraints {
     return result;
   }
 
+  /// Returns a size that attempts to meet the following conditions, in order:
+  ///
+  ///  - The size must satisfy the constraints.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: these constraints

File: packages/flutter/lib/src/rendering/box.dart (hunk 2 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: Sure.

**Comment by Hixie** [comment]
LGTM

**Pull request merged**